### PR TITLE
Migrate S3579 to eslint frontend for JS/TS (new rule for TS)

### LINKED ISSUE #2204: Migrate S3579 to eslint frontend for JS/TS (new rule for TS)

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**
Fixes #2204

### CODE REVIEW

File: eslint-bridge/src/rules/no-associative-arrays.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+// https://jira.sonarsource.com/browse/RSPEC-4619
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/src/rules/no-associative-arrays.ts:20]: Probably copy-pasted, it's not the right link.

File: eslint-bridge/src/rules/no-associative-arrays.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    function isString(node: estree.Node) {
+      const typ = checker.getTypeAtLocation(
+        services.esTreeNodeToTSNodeMap.get(node as TSESTree.Node),
+      );
+      return (typ.getFlags() & ts.TypeFlags.StringLike) !== 0;
+    }
+
+    return {
+      'AssignmentExpression[left.type="MemberExpression"]'(node: estree.Node) {
+        const memberExpr = (node as estree.AssignmentExpression).left as estree.MemberExpression;
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/src/rules/no-associative-arrays.ts:56]: No need to change anything here, but you could use destructuring here, i.e., `const { object, property } = ...;` .

File: eslint-bridge/src/rules/no-associative-arrays.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    const checker = services.program.getTypeChecker();
+
+    function isArray(node: estree.Node) {
+      const typ = checker.getTypeAtLocation(
+        services.esTreeNodeToTSNodeMap.get(node as TSESTree.Node),
+      );
+      return typ.symbol && typ.symbol.name === 'Array';
+    }
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/src/rules/no-associative-arrays.ts:45]: There is a function for that in `rules/utils.ts`.

File: eslint-bridge/src/rules/no-associative-arrays.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    function isString(node: estree.Node) {
+      const typ = checker.getTypeAtLocation(
+        services.esTreeNodeToTSNodeMap.get(node as TSESTree.Node),
+      );
+      return (typ.getFlags() & ts.TypeFlags.StringLike) !== 0;
+    }
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/src/rules/no-associative-arrays.ts:52]: Still in `rules/utils`, there is a function for that, but I think your implementation is better. Could you replace it with this piece of code ?

**Review by yassin-kammoun-sonarsource — CHANGES_REQUESTED** [review]

File: eslint-bridge/src/rules/no-associative-arrays.ts (hunk 1 above)
Comment (andrea-guarino-sonarsource) [reply eslint-bridge/src/rules/no-associative-arrays.ts:20]: Indeed, thanks!

**Review by yassin-kammoun-sonarsource — APPROVED** [review]

**Pull request merged**